File: js/src/swmmrs/exceptions.ts
/** Native identity attached to a configuration diagnostic, not a lookup handle. */
export interface ConfigurationObjectIdentity {
  /** Native object-family name. */
  readonly objectType: string;
  /** Canonical object ID. */
  readonly id: string;
  /** Native configured-order index for diagnostics. */
  readonly index: number;
}
/** One rejected configuration rule; the complete patch remains unapplied. */
export interface ConfigurationDiagnostic {
  /** Object whose declaration failed validation. */
  readonly object: ConfigurationObjectIdentity;
  /** Property path identifying the rejected value. */
  readonly propertyPath: string;
  /** Stable native validation-rule identifier. */
  readonly ruleCode: string;
  /** Human-readable explanation. */
  readonly message: string;
  /** Other object involved in the conflict, or null. */
  readonly conflictingObject: ConfigurationObjectIdentity | null;
}

/** Details retained when an error crosses the worker boundary. */
export interface ErrorDetails {
  /** Required human-readable error text. */
  message: string;
  /** Native numeric code, when supplied. */
  code?: number | undefined;
  /** Operation that failed. */
  operation?: string | undefined;
  /** Additional native detail. */
  detail?: string | undefined;
  /** Semantic category, preserving unknown future values. */
  semanticCode?: string | undefined;
  /** Native failure category. */
  category?: string | undefined;
  /** Configuration diagnostics; presence takes precedence over numeric-code mapping. */
  diagnostics?: readonly ConfigurationDiagnostic[] | undefined;
  /** Partial report text, when available. */
  report?: string | undefined;
  /** Serialized secondary cleanup failure, when supplied by the worker. */
  cleanupError?: string | undefined;
}

/** Base runtime exception for binding and solver failures; supports `instanceof`.
 * Ordinary Error message, name, and stack are retained. Standalone output failures
 * use `OutputError`, not this hierarchy.
 */
export class SwmmError extends Error {
  /** Native numeric failure code, when available. */
  readonly code: number | undefined;
  /** Failed public operation, when available. */
  readonly operation: string | undefined;
  /** Additional native failure detail. */
  readonly detail: string | undefined;
  /** Preserves unknown future categories rather than discarding them. */
  readonly semanticCode: string | undefined;
  /** Native error category. */
  readonly category: string | undefined;
  /** Partial report text, when available. */
  readonly report: string | undefined;
  /** Secondary cleanup failure retained after a primary run failure. */
  cleanupError: unknown;

  /** Construct an error with serialized or caller-supplied details.
   * @param details - Message and optional structured failure metadata.
   */
  constructor(details: ErrorDetails) {
    super(details.message);
    this.name = new.target.name;
    this.code = details.code;
    this.operation = details.operation;
    this.detail = details.detail;
    this.semanticCode = details.semanticCode;
    this.category = details.category;
    this.report = details.report;
    this.cleanupError = details.cleanupError;
  }
}
/** Native failure without a more specific binding category, including unmapped future codes. */
export class SolverError extends SwmmError {}
/** Invalid phase, closed owner, or competing advancement. Native codes: 501, 502, 503, 2001, 2002, 2013. */
export class LifecycleError extends SwmmError {}
/** Retained owner-bound view no longer matches the live owner. */
export class StaleViewError extends LifecycleError {}
/** Invalid value or read. Native codes: 504, 507, 508, 509, 2003, 2006, 2010, 2012. */
export class ValidationError extends SwmmError {}
/** Configuration rejected with object-level diagnostics; takes precedence over code-based categorization. */
export class ConfigurationError extends ValidationError {
  /** Frozen diagnostic array with frozen object identities. */
  readonly diagnostics: readonly ConfigurationDiagnostic[];
  /** Construct a configuration failure and defensively freeze its diagnostics.
   * @param details - Error metadata and optional diagnostic list (defaults to empty).
   */
  constructor(details: ErrorDetails) {
    super(details);
    this.diagnostics = Object.freeze((details.diagnostics ?? []).map(diagnostic => Object.freeze({
      ...diagnostic,
      object: Object.freeze({ ...diagnostic.object }),
      conflictingObject: diagnostic.conflictingObject === null ? null : Object.freeze({ ...diagnostic.conflictingObject }),
    })));
  }
}
/** Unknown object/definition ID. Native codes: 505, 506, 2000, 2004, 2005, 2007, 2008, 2009. */
export class ObjectNotFoundError extends SwmmError {}
/** Worker startup, communication, or runtime prerequisite failure. */
export class WorkerError extends SwmmError {}
/** Internal native simulation failure, mapped from code 9999. */
export class InternalSimulationError extends WorkerError {}

/** @internal */
export function fromWorkerError(details: ErrorDetails): SwmmError {
  if (details.diagnostics !== undefined) return new ConfigurationError(details);
  if ([501, 502, 503, 2001, 2002, 2013].includes(details.code ?? -1)) return new LifecycleError(details);
  if ([505, 506, 2000, 2004, 2005, 2007, 2008, 2009].includes(details.code ?? -1)) return new ObjectNotFoundError(details);
  if ([504, 507, 508, 509, 2003, 2006, 2010, 2012].includes(details.code ?? -1)) return new ValidationError(details);
  if (details.code === 9999) return new InternalSimulationError(details);
  return new SolverError(details);
}

File: python/src/swmmrs/exceptions.py
"""Public exception hierarchy for Python simulation operations."""

from __future__ import annotations as _annotations

from dataclasses import dataclass as _dataclass
from typing import final as _final

from ._prettier import pretty_dataclass as _pretty_dataclass
from .enums import SolverErrorCode as _SolverErrorCode

_STANDARD_EXCEPTION_ATTRIBUTES = frozenset(
    (
        "args",
        "__cause__",
        "__context__",
        "__notes__",
        "__suppress_context__",
        "__traceback__",
    )
)


class SwmmError(Exception):
    """Base class for supported package failures.

    Notes
    -----
    Catch this class to handle all public binding failures without catching
    unrelated Python exceptions.
    """

    __slots__ = ("native_code", "operation", "detail", "semantic_code")

    native_code: int | None
    operation: str | None
    detail: str | None
    semantic_code: str | None

    def __init__(
        self,
        message: str = "",
        native_code: int | None = None,
        operation: str | None = None,
        detail: str | None = None,
        semantic_code: str | None = None,
    ) -> None:
        self.native_code = native_code
        self.operation = operation
        self.detail = detail
        self.semantic_code = semantic_code
        super().__init__(message)

    def __setattr__(self, name: str, value: object) -> None:
        slots = {
            slot
            for class_ in type(self).__mro__
            for slot in class_.__dict__.get("__slots__", ())
        }
        if name not in _STANDARD_EXCEPTION_ATTRIBUTES and name not in slots:
            raise AttributeError(f"{type(self).__name__} has no attribute {name!r}")
        super().__setattr__(name, value)


class LifecycleError(SwmmError):
    """Report an operation that is invalid in the current lifecycle state."""

    __slots__ = ()


@_final
class StaleViewError(LifecycleError):
    """Report access through a Live View from an inactive Project Generation."""

    __slots__ = ()


@_final
class ValidationError(SwmmError):
    """Report a public argument rejected before native state mutation."""

    __slots__ = ()


@_pretty_dataclass(frozen=True, slots=True)
class ConfigurationObjectIdentity:
    """Identify one configured object named by a configuration diagnostic."""

    object_type: str
    id: str
    index: int


@_pretty_dataclass(frozen=True, slots=True)
class ConfigurationDiagnostic:
    """Describe one ordered post-open configuration violation."""

    object: ConfigurationObjectIdentity
    property_path: str
    rule_code: str
    message: str
    conflicting_object: ConfigurationObjectIdentity | None


@_final
class ConfigurationError(SwmmError):
    """Report post-open preparation diagnostics without invalidating the owner."""

    __slots__ = ("diagnostics",)

    diagnostics: tuple[ConfigurationDiagnostic, ...]

    def __init__(self, diagnostics: tuple[ConfigurationDiagnostic, ...]) -> None:
        self.diagnostics = diagnostics
        super().__init__(
            f"start: post-open preparation rejected {len(diagnostics)} configuration violation(s)"
        )

    def __str__(self) -> str:
        """Include every ordered structured diagnostic in traceback text."""

        summary = super().__str__()
        if not self.diagnostics:
            return summary
        lines = [summary + ":"]
        for diagnostic in self.diagnostics:
            identity = diagnostic.object
            line = (
                f"- {identity.object_type} {identity.id!r} (index {identity.index}), "
                f"property {diagnostic.property_path!r}, rule {diagnostic.rule_code!r}: "
                f"{diagnostic.message}"
            )
            if diagnostic.conflicting_object is not None:
                conflicting = diagnostic.conflicting_object
                line += (
                    f"; conflicts with {conflicting.object_type} {conflicting.id!r} "
                    f"(index {conflicting.index})"
                )
            lines.append(line)
        return "\n".join(lines)


@_final
class SolverError(SwmmError):
    """Report a project or lifecycle operation rejected by the native solver.

    Attributes
    ----------
    code : SolverErrorCode
        Stable category for the native failure.
    native_code : int
        Numeric SWMM error code.
    operation : str
        Public operation that triggered the failure.
    detail : str or None
        Optional solver-provided diagnostic detail, including any secondary
        cleanup failure retained by the solver.
    semantic_code : str
        Stable semantic failure code.
    """

    __slots__ = ("code",)

    code: _SolverErrorCode
    operation: str
    detail: str | None
    semantic_code: str

    def __init__(
        self,
        code: _SolverErrorCode,
        operation: str,
        detail: str | None = None,
        native_code: int | None = None,
        semantic_code: str | None = None,
    ) -> None:
        self.code = code
        semantic_code = semantic_code or code.value
        message = f"{operation}: {code.message}"
        super().__init__(
            message if detail is None else f"{message}: {detail}",
            native_code,
            operation,
            detail,
            semantic_code,
        )

@_final
class InternalSimulationError(SwmmError):
    """Report an isolated native Simulation Owner that is permanently unavailable."""

    __slots__ = ()


__all__ = (
    "ConfigurationDiagnostic",
    "ConfigurationError",
    "ConfigurationObjectIdentity",
    "InternalSimulationError",
    "LifecycleError",
    "SolverError",
    "StaleViewError",
    "SwmmError",
    "ValidationError",
)
